application/job/job.go

```diff
@@ -0,0 +1,100 @@
+package job
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gobuffalo/buffalo/worker"
+	"github.com/gobuffalo/envy"
+	"github.com/silinternational/wecarry-api/domain"
+	"github.com/silinternational/wecarry-api/models"
+	"github.com/silinternational/wecarry-api/notifications"
+)
+
+const (
+	NewMessage = "new_message"
+)
+
+var W worker.Worker
+
+func init() {
+	W = worker.NewSimple()
+	if err := W.Register(NewMessage, NewMessageHandler); err != nil {
+		domain.ErrLogger.Printf("error registering 'new_message' worker, %s", err)
+	}
+}
+
+// NewMessageHandler is the Worker handler for new notifications of new Thread Messages
+func NewMessageHandler(args worker.Args) error {
+	id, ok := args[domain.ArgMessageID].(int)
+	if !ok {
+		return fmt.Errorf("no message ID provided to new_message worker, args = %+v", args)
+	}
+
+	var m models.Message
+	if err := m.FindByID(id, "SentBy", "Thread"); err != nil {
+		return fmt.Errorf("bad ID (%d) received by new message handler, %s", id, err)
+	}
+
+	if err := m.Thread.Load("Participants", "Post"); err != nil {
+		return errors.New("failed to load Participants and Post in new message handler")
+	}
+
+	uiUrl := envy.Get(domain.UIURLEnv, "")
+	msg := notifications.Message{
+		Template: domain.MessageTemplateNewMessage,
+		Data: map[string]interface{}{
+			"postURL":        uiUrl + "/#/requests/" + m.Thread.Post.Uuid.String(),
+			"postTitle":      m.Thread.Post.Title,
+			"messageContent": m.Content,
+			"sentByNickname": m.SentBy.Nickname,
+			"threadURL":      uiUrl + "/#/messages/" + m.Thread.Uuid.String(),
+		},
+		FromName:  m.SentBy.Nickname,
+		FromEmail: m.SentBy.Email,
+	}
+
+	for _, p := range m.Thread.Participants {
+		if p.ID == m.SentBy.ID {
+			continue
+		}
+
+		var tp models.ThreadParticipant
+		if err := tp.FindByThreadIDAndUserID(m.ThreadID, p.ID); err != nil {
+			return err
+		}
+		// Don't send a notification if this user has viewed the message or if they've already been notified
+		if tp.LastViewedAt.After(m.UpdatedAt) || tp.LastNotifiedAt.After(m.UpdatedAt) {
```

Comment:
Does`tp.LastNotifiedAt.After(m.UpdatedAt)` mean if a notification was sent to me yesterday, and a new message is posted in the thread today that I will not receive a new notification today because I was notified yesterday?

Reply:
No, this is looking at the `updated_at` for the message that triggered the event. So this is only preventing a second notification for a given message. In other words, if two messages have been posted since the last notification, only the first will trigger a notification.